### PR TITLE
Make #16284 backward compatible

### PR DESCRIPTION
**Pull request by adithyachakilam**
In #16284, we changed the way how kinesis autoscaler computes the lag for the purposes of autoscaling. Put that feature behind a flag so as to not annoy customers who have already configured the auto scaler thesholds based on the total lag.

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/LagBasedAutoScalerConfig.java
Comment (suneet-s) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/LagBasedAutoScalerConfig.java:66]: I think it would be better to have this be an enum, like what was done in the initial patch. Then the docs for this field would tell the user to set this to one of `TOTAL`, `MAX` or `AVG` with the default implementation using the total lag.

@kfaraz do you have any concerns with this approach since you had some comments about the use of an enum in an earlier patch.

On naming - I think the name could be something like `lagStatsType` The word `default` in the current name is confusing to me. Once there is agreement on the approach, there should also be a doc update in the `supervisor.md` page for this new config option.

Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/LagBasedAutoScalerConfig.java:66]: I agree that the name should be `lagStatsType` or if you just want to use max, then `useMaxLag`. I don't mind it being an enum, my only concern is that it is limiting. What do we do if we want something outside of total/max/avg?

The approach in this PR is not exactly the same as the original PR, because here we drive the decision by a config and not a method on the `Supervisor` itself.

I would prefer it we could make the following changes in this PR:
1. If we intend to drive the decision via a config, let's get rid of the method `computeLagForAutoScaler()` altogether.
2. Update `LagStats` (returned by `computeLagStats()`) to have a field `preferredScalingMetric` which would be the new enum we add. For now, this value will be `TOTAL` for both Kinesis and Kafka. After we have done some validation, the `KinesisSupervisor.computeLagStats()` would return a `LagStats` with `preferredScalingMetric` as `MAX` for Kinesis.
3. The new config will essentially override the `preferredScalingMetric`. If the config is not specified, we use the `preferredScalingMetric` returned by the `computeLagStats` method. This logic should be in the `LagBasedAutoScaler` itself.
4. Let's have a plan on how we will eventually get rid of the new config or atleast compute its default value better. Because a single default value of this config is not going to work for all kinds of supervisors. We wouldn't want users to have to specify this config in every supervisor spec. I think step 2 above will help with this.

---

My concerns with the original approach were:
1. It was computing the lag in a roundabout manner. Supervisor needed to expose two methods `computeLagStats` and `getLagMetric`. Then the auto-scaler had to use the result of these two methods to determine the actual lag. If it is the Supervisor which drives this computation, the Supervisor might as well have a single method that returns the required lag value.
2. Enum is limited to TOTAL, MAX, AVG. Having a separate method for `computeLagForAutoScaler` allows us to keep the implementation more flexible. I don't know when we would need a value outside of total / max / avg though, perhaps if we wanted a normalized / truncated value of one of these metrics.
3. I didn't fully like the idea of adding a new method `computeLagForAutoScaler` either because the method `computeLagStats` itself was meant only to be used by the auto-scaler. Why have one more method? I suggested that approach as I didn't have a better alternative at that time which could move things along faster.

Comment (suneet-s) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/LagBasedAutoScalerConfig.java]: @kfaraz Cool, makes sense to me.

File: server/src/main/java/org/apache/druid/indexing/overlord/supervisor/autoscaler/ScalingMetric.java
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/indexing/overlord/supervisor/autoscaler/ScalingMetric.java:22]: I think a better name would be `AggregateFunction`.

File: server/src/main/java/org/apache/druid/indexing/overlord/supervisor/autoscaler/ScalingMetric.java
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/indexing/overlord/supervisor/autoscaler/ScalingMetric.java:25]:
```suggestion
  SUM,
```

File: server/src/main/java/org/apache/druid/indexing/overlord/supervisor/autoscaler/LagStats.java
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/indexing/overlord/supervisor/autoscaler/LagStats.java:27]: Nit: Thinking about the name again, we could call this field `preferredAggregateForScaling` or even simply `aggregateForScaling`.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/LagBasedAutoScalerConfig.java
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/LagBasedAutoScalerConfig.java:67]: Suggestion for rename:
```suggestion
          @Nullable @JsonProperty("lagAggregate") AggregateFunction lagAggregate
```

**Review by kfaraz — COMMENTED** [review]
@adithyachakilam , thanks for the changes! The implementation makes sense. I have just left some suggestions for field names and code structure.

**Review by kfaraz — COMMENTED** [review]
Final comments, will approve once these are addressed.

File: indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamSupervisorSpecTest.java
Comment (adithyachakilam) [reply indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamSupervisorSpecTest.java:1339]: added verifications.

File: indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamSupervisorSpecTest.java
Comment (kfaraz) [inline indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamSupervisorSpecTest.java:138]: `LagStats` is a simple POJO. We shouldn't have to mock it.

File: server/src/main/java/org/apache/druid/indexing/overlord/supervisor/autoscaler/LagStats.java
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/indexing/overlord/supervisor/autoscaler/LagStats.java:63]: This is not nullable (see constructor of this class). `LagStatsAutoScalerConfig.getLagAggregate()` is nullable.

File: indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamSupervisorSpecTest.java
Comment (kfaraz) [inline indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamSupervisorSpecTest.java:152]: These are weird expectations. Why do we care how many times sum and max were called?

File: indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamSupervisorSpecTest.java
Comment (kfaraz) [inline indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamSupervisorSpecTest.java:1011]: Instead of these assertions, we just need a test in `LagStatsTest` that verifies the results of the method `LagStats.getMetric()`.

Comment (adithyachakilam) [reply indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamSupervisorSpecTest.java:1011]: I think that test is already there now. This test verifies the override by the config is working fine.

File: indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamSupervisorSpecTest.java
Comment (adithyachakilam) [reply indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamSupervisorSpecTest.java:138]: I need to spy that class so as to confirm that the config override is working and not calling SUM when MAX is specified and viceversa.

File: indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamSupervisorSpecTest.java
Comment (kfaraz) [reply indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamSupervisorSpecTest.java:1011]: > I think that test is already there now.
Thanks, I must have missed that.

Comment (kfaraz) [reply indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamSupervisorSpecTest.java:1011]: > This test verifies the override by the config is working fine.

The correct way to verify that would be to have different `LagAutoScalerConfig`s and different supervisor impls (which return different values of `LagStats.aggregateForScaling`)  and verify that the value of `taskCountAfterScaleOut` changes based on whether the config/supervisor uses MAX or SUM or AVERAGE.
(These tests can also be added later and need not block this PR, unless required for code coverage).

We should not have to mock or spy the `LagStats` class just to check which method was invoked on it. For this test class, the end result is the task scale out and that is all we need to verify. We cannot verify every internal method called when invoking any public interface.

Please revert these test changes, rest of the PR looks okay to me.

**Review by kfaraz — APPROVED** [review]
Thanks for addressing the comments, @adithyachakilam .

**Comment by kfaraz** [comment]
Merging this PR as the pending tests are stuck due to the issue described in #16347 . The same tests have already passed on a different JDK.

**Comment by suneet-s** [comment]
@adithyachakilam Can you please backport this change to the 30.0 branch as it looks like #16284 is in that branch

**Pull request merged**
